### PR TITLE
Cleanup a number of warnings in the code.

### PR DESCRIPTION
**Pull request by rsgowman**
Test code still has warnings; those aren't addressed here.

* Eliminate 'badimport' warning

  It was already reasonably obvious what this is, so the warning is mostly
  spurious. But it's easier and less boilerplate to just add the extra
  qualifier than to suppress the warning.

  https://errorprone.info/bugpattern/BadImport

* Suppress EqualsGetClass warnings for public classes

  These are effectively final and are documented as not being final only
  to allow mocking within tests.

  https://errorprone.info/bugpattern/EqualsGetClass

* Mark a number of classes final

  Mostly to avoid a warning, though it's generally recommended to do this
  anyways. (Note that this makes mocking impossible; if any of these
  classes need to be mocked, we might be better of just suppressing the
  warning or using instanceof.)

  https://errorprone.info/bugpattern/EqualsGetClass

* Switch .getClass() to instanceof in some .equals() methods

  This is for cases where we're dealing with a non-leaf node in the
  inheritance hierarchy. (Leaf nodes in the hierarchy can just mark the
  entire class as final.) In order to preserve symmetry of .equals(), also
  mark the .equals() methods themselves as final.

  In this case, "non-leaf" nodes also counts effectively final classes
  which are mocked. (i.e. ViewSnapshot.)

  https://errorprone.info/bugpattern/EqualsGetClass

* Provide toString on FieldMask

  Since it's used by PatchMutation

  https://errorprone.info/bugpattern/ObjectToString

* Fix a 'missing override' warning

  https://errorprone.info/bugpattern/MissingOverride

* Rework FieldMask to use a Set of FieldPaths rather than a Collection.

  FieldMask uses .equals() on the container, and that's not well defined
  for a Collection. The fix is to switch to either a Set or a List. Set
  seems more appropriate.

  https://errorprone.info/bugpattern/UndefinedEquals

### CODE REVIEW

**Comment by rsgowman** [comment]
Note that each commit within this PR is more-or-less independent of the others. If anything's controversial, we can just kick it out and handle it separately.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/FieldMask.java [hunk 1]
```diff
@@ -26,13 +26,13 @@
  * object foo. If foo is not an object, foo is replaced with an object containing foo.
  */
 public final class FieldMask {
-  public static FieldMask fromCollection(Collection<FieldPath> mask) {
+  public static FieldMask fromSet(Set<FieldPath> mask) {
     return new FieldMask(mask);
   }
 
-  private final Collection<FieldPath> mask;
+  private final Set<FieldPath> mask;
```
Comment (var-const) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/FieldMask.java]: Optional: what's the reason to prefer `Set` to a `List`? AFAICS, `mask` is never used for lookup; the only access is iterating over all the elements (line 63), for which `List` will likely be faster.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentReference.java [hunk 2]
```diff
@@ -525,6 +525,7 @@ private ListenerRegistration addSnapshotListenerInternal(
   }
 
   @Override
+  @SuppressWarnings("EqualsGetClass") // This class is effectively final.
```
Comment (var-const) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentReference.java]: Question: why not use `instanceof` here and in `Query#equals`? The article you linked states that `instanceof` is never worse than `getClass`:

> What about final or effectively-final classes?
> 
> In this case there is no disadvantage to the `getClass` trick - but there’s no great advantage to it either. Most unsafe idioms have circumstances in which they are safe, but this doesn’t change the fact that they are generally unsafe and not worth propagating and legitimizing.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/ArrayTransformOperation.java [hunk 3]
```diff
@@ -51,11 +51,11 @@ public FieldValue applyToRemoteDocument(FieldValue previousValue, FieldValue tra
   }
 
   @Override
-  public boolean equals(Object o) {
+  public final boolean equals(Object o) {
     if (this == o) {
       return true;
     }
-    if (o == null || getClass() != o.getClass()) {
+    if (!(o instanceof ArrayTransformOperation)) {
```
Comment (mikelehen) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/ArrayTransformOperation.java]: Hrm.  I'm not sure about this...  We have two implementations of ArrayTransformOperation: ArrayTransformOperation.Union, and ArrayTransformOperation.Remove.  They shouldn't be considered equal even if they have the same elements being unioned or removed.

I'm not sure what this means.  Honestly, getClass() equality is exactly what we want, and I'm not sure how much I care about the substitutability principle...  So we could keep the existing code but suppress the warning...  or we could override equals again in the Union / Remove classes to do the right thing.  WDYT?

File: firebase-firestore/src/testUtil/java/com/google/firebase/firestore/testutil/TestUtil.java [hunk 4]
```diff
@@ -431,12 +433,12 @@ public static PatchMutation patchMutation(
     boolean merge = updateMask != null;
 
     // We sort the fieldMaskPaths to make the order deterministic in tests.
-    Collections.sort(objectMask);
+    SortedSet<FieldPath> fieldMaskPaths = new TreeSet<>(merge ? updateMask : objectMask);
```
Comment (mikelehen) [inline firebase-firestore/src/testUtil/java/com/google/firebase/firestore/testutil/TestUtil.java]: Hrm.  I wonder if this is really necessary.  Were we just sorting them before so equals() would work out?  If so, then we're already good.  What happens if you use a HashSet here?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/FieldMask.java (hunk 1 above)
Comment (mikelehen) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/FieldMask.java]: I agree that Set is the natural data structure (and perf, etc. aren't major concerns).  That said, we may now get slightly different behavior than on other platforms (e.g. with regard to duplicates), though I don't think that's a big deal, and the Android behavior will probably be more correct.

File: firebase-firestore/src/testUtil/java/com/google/firebase/firestore/testutil/TestUtil.java (hunk 4 above)
Comment (mikelehen) [reply firebase-firestore/src/testUtil/java/com/google/firebase/firestore/testutil/TestUtil.java]: Oh. I thought it was a HashSet.  In any case, I'm fine with using a TreeSet, but the comment about sorting seems confusing / wrong now.  Perhaps just remove it?

Comment (mikelehen) [reply firebase-firestore/src/testUtil/java/com/google/firebase/firestore/testutil/TestUtil.java]: Ah! Comment makes perfect sense now, thanks.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/FieldMask.java (hunk 1 above)
Comment (mikelehen) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/FieldMask.java]: I'm fine with any of these options (including keeping it as a Set as you have it now).

**Pull request merged**